backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objc/linkObjC.kt

```diff
@@ -0,0 +1,228 @@
+
+        val globalPatch = nameToGlobalPatch[name]
+        if (globalPatch != null) {
+            LLVMSetValueName(global, globalPatch.newGlobalName)
+            unusedPatches -= globalPatch
+        } else if (PatchBuilder.GlobalKind.values().any { name.startsWith(it.prefix) }) {
+            error("Objective-C global '$name' is not patched")
+        }
+
+        val section = LLVMGetSection(global)?.toKString()
+        sectionToValueToLiteralPatch[section]?.let { valueToLiteralPatch ->
+            val value = getStringValue(initializer)
+            val patch = valueToLiteralPatch[value]
+            if (patch != null) {
+                if (patch.newValue != value) patchLiteral(global, patch.generator, patch.newValue)
+                unusedPatches -= patch
+            } else if (section == ObjCDataGenerator.classNameGenerator.section) {
+                error("Objective-C class name literal is not patched: $value")
+            }
+        }
+    }
+
+    unusedPatches.firstOrNull()?.let {
+        error("Patch is not applied: $it")
+    }
+}
+
+private fun getStringValue(initializer: LLVMValueRef): String? = when (LLVMGetValueKind(initializer)) {
+    LLVMValueKind.LLVMConstantDataArrayValueKind -> memScoped {
+        require(LLVMIsConstantString(initializer) != 0) { "not a constant string: ${llvm2string(initializer)}" }
+
+        val lengthVar = alloc<size_tVar>()
+        val bytePtr = LLVMGetAsString(initializer, lengthVar.ptr)!!
+        val length = lengthVar.value
+
+        val lastByte = bytePtr[length - 1]
+        require(lastByte == 0.toByte()) {
+            "${llvm2string(initializer)}:\n  expected zero terminator, found $lastByte"
+        }
+
+        bytePtr.toKString()
+    }
+
+    LLVMValueKind.LLVMConstantAggregateZeroValueKind -> ""
+
+    else -> error("Unexpected literal initializer: ${llvm2string(initializer)}")
+}
+
+private fun <T, K> List<T>.associateNonRepeatingBy(keySelector: (T) -> K): Map<K, T> =
+        this.groupBy(keySelector)
+                .mapValues { (key, values) ->
+                    values.singleOrNull()
+                            ?: error("multiple values found for $key: ${values.joinToString()}")
+                }
+
+private fun patchLiteral(
+        global: LLVMValueRef,
+        generator: ObjCDataGenerator.CStringLiteralsGenerator,
+        newValue: String
+) {
+    val module = LLVMGetGlobalParent(global)!!
+
+    val newFirstCharPtr = generator.generate(module, newValue).getElementPtr(0).llvm
+
+    generateSequence(LLVMGetFirstUse(global), { LLVMGetNextUse(it) }).forEach { use ->
+        val firstCharPtr = LLVMGetUser(use)!!.also {
+            require(it.isFirstCharPtr(global)) {
+                "Unexpected literal usage: ${llvm2string(it)}"
+            }
+        }
+        LLVMReplaceAllUsesWith(firstCharPtr, newFirstCharPtr)
+    }
+}
+
+private fun LLVMValueRef.isFirstCharPtr(global: LLVMValueRef): Boolean =
```

Comment:
Looks fragile :(.

Reply:
The purpose of this predicate is to help detecting changes in clang-produced Objective-C bitcode.
This pattern hasn't been changed for a long time, and I don't expect it going to.